Add user interface to search through a publication

gatamar opened the pull request:
This PR aims to add a Search UI for the Swift Test App: https://github.com/readium/r2-testapp-swift/issues/386
The implementation is based on https://github.com/readium/r2-testapp-swift/discussions/402

I'm open to any feedback/suggestions of how to improve it (it's far from perfect, but I'd like to hear any UI/UX suggestions before doing an extra work).

Particularly, I'm not sure in such things:
1. How to display the search results in a search table? Can it be like now (`UITextView` + `NSAttributedString`), or some `WKWebView` + `Decoration` would be better (and that there's some API for this)?
2. When we tap on a result, we navigate to it's location in text, and a highlight is added. Should this highlight be cleared when a user taps somewhere else?
3. Currently the search query is cached in a ViewModel, such that when go to search the next time, it's still there(like in Apple Books). Should an index in a result list be also cached?

A video of how it works: https://drive.google.com/file/d/1hFzTsy5DpTXIWBJqFJ6tny1JNIcnxGcL/view?usp=sharing

mickael-menu commented:
Thanks for taking the time to implement this!

I'm going to review the code on Friday, but here are some answers to your questions.

> * How to display the search results in a search table? Can it be like now (`UITextView` + `NSAttributedString`), or some `WKWebView` + `Decoration` would be better (and that there's some API for this)?

`UITextView` is fine, especially because the `locator.text` doesn't contain any HTML so the `WKWebView` would not help much there.

If you prefer, you can also implement it with SwiftUI. I'm planning on migrating the test app to SwiftUI progressively. That's not required for your PR though.

> * When we tap on a result, we navigate to it's location in text, and a highlight is added. Should this highlight be cleared when a user taps somewhere else?

I guess each app might do things differently, but to keep things simple in the test app, maybe you can clear the highlight when you receive a new `navigator(_:locationDidChange:)` event? This way it should clear when turning pages.

> * Currently the search query is cached in a ViewModel, such that when go to search the next time, it's still there(like in Apple Books). Should an index in a result list be also cached?

Caching the search query is good. By index in a result list, you mean the scroll position? If you can handle this without too much code that would be useful. The test app doesn't need to be perfect but just show how to use the Readium API.

gatamar commented:
@mickael-menu  Thanks for the feedback!

> you can also implement it with SwiftUI. I'm planning on migrating the test app to SwiftUI progressively

I believe SwiftUI is a great idea, updated this PR to SwiftUI (and both UI/UX became nicer 🙂). 
However, when trying to use some not-so-new SwiftUI API (e.g. `ScrollViewReader`, `LazyVStack`), a preprocessor complains that it's available from _iOS 14_, while the project uses _iOS 13.6_. Do I understand correctly that updating the project's Deployment target isn't a problem, because it's a Sample app, so no one depends on it?

---

> By index in a result list, you mean the scroll position?

Thanks, I meant exactly that. Wasn't able to do it quickly, can try again after hearing you feedback about updating Deployment target (the chances are high that geting/setting the lists's scroll position will require iOS 14 API).

---

> maybe you can clear the highlight when you receive a new `navigator(_:locationDidChange:)` event

I tried this, and it doesn't seem to work in `ReaderViewController`, because this event happens to be called a lot of times, so the highlight is cleared even before it's added! The steps are following:
1. The user selects some locator on a Search screen
2. "navigator.go(to: locator)" is called
3. only in the  completion handler of "navigator.go(to: locator)" a highlight is added
4. but some callbacks(of the navigation for which we just received the callback?) are triggered, and `navigator(_:locationDidChange:)` is called, so the selection is cleared before the user saw it.

As you advised to clear the selection on the page change, maybe it should be in `EPUBNavigationViewController`?
```
/// Goes to the next or previous page in the given scroll direction.
private func go(to direction: EPUBSpreadView.Direction, animated: Bool, completion completionBlock: @escaping () -> Void) -> Bool
```
But as `EPUBNavigationViewController` is inside `R2Navigator` package, and the app uses a particular version `2.2.0`, I don't know how this supposed to be handled in terms of dependency management/PRs/blockers.

mickael-menu commented:
I'll take a look on Friday what can be done for clearing the search term.

> Do I understand correctly that updating the project's Deployment target isn't a problem, because it's a Sample app, so no one depends on it?

Yeah that's fine. I think for the test app we can follow Apple's best practices of supporting the last two iOS versions.

mickael-menu commented on TestApp/Sources/Reader/Common/Search/SearchView.swift:
```diff
@@ -0,0 +1,95 @@
+}
+
+extension Locator {
+    /// Sometimes when there's an image before the text, "text.before" looks like "\n\t\t\n\t\n\n\t\n\n\t Some Seach word"
+    /// This function tries to remove leading tabs and newlines for a more compact preview
+    var previewTextBefore: String {
+        let textBefore = text.before ?? ""
+        return String(textBefore.drop { char in char == "\n" || char == "\t" })
+    }
+    
+    var previewTextHighlight: String {
+        return text.highlight ?? ""
+    }
+    
+    var previewTextAfter: String {
+        return String(text.after ?? "")
+    }
```
Yeah the `Locator.Text` is not sanitized (closest to the HTML content), so you might get some extra whitespace characters. But this could be also inside the `highlight` and `after` as well as between two words.

I suggest adding a String extension to coalesce any whitespaces into a single space and use it on all `Locator.Text` properties:

```swift
extension String {
    /// Replaces multiple whitespaces by a single space.
    func coalescingWhitespaces() -> String {
        return replacingOccurrences(of: "[\\s\n]+", with: " ", options: .regularExpression, range: nil)
            .trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
    }
}

extension Locator.Text {
    var previewBefore: String {
        before?.coalescingWhitespaces()?.removingPrefix(" ") ?? ""
    }

    var previewHighlight: String {
        highlight?.coalescingWhitespaces() ?? ""
    }

    var previewAfter: String {
        after?.coalescingWhitespaces() ?? ""
    }
}
```

mickael-menu commented on TestApp/Sources/Reader/Common/ReaderViewController.swift:
```diff
@@ -184,6 +192,30 @@ class ReaderViewController: UIViewController, Loggable {
+                    //self.dismiss(animated: false, completion: nil)
+                    self.navigator.go(to: locator, animated: true) {
+                        if let decorator = self.navigator as? DecorableNavigator {
+                            let decoration = Decoration(id: locator.text.jsonString ?? "", locator: locator, style: Decoration.Style.highlight(tint: .yellow, isActive: false))
```
You can use the same ID for the `Decoration` (e.g. `selectedSearchResult`), as we display only one.

mickael-menu commented on TestApp/Sources/Reader/Common/Search/SearchView.swift:
```diff
@@ -0,0 +1,93 @@
+
+struct SearchView: View {
+    @ObservedObject var viewModel: SearchViewModel
+    @State var query: String = ""
+    var body: some View {
+        let queryValueBinding = Binding<String>(get: {
+            self.query
+        }, set: {
+            self.query = $0
+            viewModel.search(with: query)
+        })
```
I think you can simplify this part by exposing a `@Published private(set) var query` in the view model, which will be set by `viewModel.search(with:)`. This way you don't need the local `query` property and the last query will be used when reopening the `SearchView` as the view model is cached.

mickael-menu reviewed (changes requested):
I took a look for clearing the search highlight. Looks like it's going to be complicated with `locationDidChange:` so let's do your initial idea of clearing the highlight when the user taps the screen (`didTapAt:`).

I pushed a commit on your repo by the way, to fix the file headers.

gatamar commented:
@mickael-menu  Thanks for your guidance! I addressed the requrested changes.

1. `coalescingWhitespaces` extension for `Locator.Text`  - added it, but made a small modification to leave the last space of `before` and the first space of `after`, such that when we search "great" in "this great app", the preview will look "this **great** app", not "this**great**app".

2. Added caching the scroll position - now when we open a SearchView the next time, the last selected cell is the top-most visible cell.

mickael-menu reviewed (approved):
Great work @gatamar, thanks!

> Btw, I have a strange compile error, as I understand, my changes are an indirect reason of it(I didn't change anything in this file)

I think it might have been caused by upgrading to iOS 14. I removed this statement as I guess with iOS 14 the failure type is inferred automatically.

I also figured that this `Locator.Text` sanitizing could be useful in Readium itself, so I moved the coalescing part into `locator.text.sanitized()`.

Your PR looks good to merge for me, but I'll let you review my changes before merging, in case you see something. Thanks again!

gatamar commented:
> Your PR looks good to merge for me, but I'll let you review my changes before merging, in case you see something. Thanks again!

@mickael-menu  Thanks! Please merge it.  
Especially thanks for the unit test - sorry, I should have added it myself for that String extension. Also as you added it, I didn't need to test on device to make sure that the space between words doesn't disappear :)
```
XCTAssertEqual(
    Locator.Text(
        after: " after selection",
        before: "before selection ",
        highlight: "current selection"
    ).sanitized(),
    Locator.Text(
        after: " after selection",
        before: "before selection ",
        highlight: "current selection"
    )
)
```
